mempool/iterators.go

```diff
@@ -116,45 +116,48 @@ func NewBlockingIterator(mem *CListMempool) Iterator {
 func (iter *BlockingIterator) WaitNextCh() <-chan Entry {
 	ch := make(chan Entry)
 	go func() {
+		// Pick a lane containing the next entry to access.
+		lane, ok := iter.pickLane()
+		if !ok {
+			// There are no transactions to take from any lane. Wait until a new
+			// transaction is added to the mempool.
+			<-iter.mp.newTxCh()
+			lane = iter.mp.latestTxLane()
```

Comment:
I have a small concern about this potentially breaking the WRR algorithm
Between lines 124 and 125, there's no lock held, so many things could happen between the two. For instance 3 transactions could be added to three different lanes. If that's the case, we'll pick the last of those three transactions (rather than following WRR).

Since anyway the callsite `WaitNextCh` accepts `nil`, I would just (1) wait on `newTxCh`, return `nil`, and let the callsite (the reactor) call us again, so we can apply WRR (and we know there's something as we unblocked, so no busy-wait).

Let me try to express it in code changes... bear with me

Reply:
Something like this:
```diff
diff --git a/mempool/iterators.go b/mempool/iterators.go
--- a/mempool/iterators.go
+++ b/mempool/iterators.go
@@ -122,7 +122,8 @@ func (iter *BlockingIterator) WaitNextCh() <-chan Entry {
 			// There are no transactions to take from any lane. Wait until a new
 			// transaction is added to the mempool.
 			<-iter.mp.newTxCh()
-			lane = iter.mp.latestTxLane()
+			close(ch)
+			return
 		}
 
 		// Add the next entry to the channel if not nil.
```

Reply:
Yeah, it makes sense. That would break WRR. How about something like the following, so we do everything inside the same function, without closing and opening the channel every time:
```diff
diff --git a/mempool/iterators.go b/mempool/iterators.go
index ead951044..db2123fc7 100644
--- a/mempool/iterators.go
+++ b/mempool/iterators.go
@@ -118,11 +118,11 @@ func (iter *BlockingIterator) WaitNextCh() <-chan Entry {
        go func() {
                // Pick a lane containing the next entry to access.
                lane, ok := iter.pickLane()
-               if !ok {
+               for !ok {
                        // There are no transactions to take from any lane. Wait until a new
                        // transaction is added to the mempool.
                        <-iter.mp.addedTxCh()
-                       lane = iter.mp.latestTxLane()
+                       lane, ok = iter.pickLane()
                }
 
                // Add the next entry to the channel if not nil.
```

Reply:
Yeah, like this one more

Reply:
Done.

Reply:
🙏